lib/configuration/variables/parse.js

```diff
@@ -0,0 +1,452 @@
+);
+const isType = RegExp.prototype.test.bind(/^[a-z][a-zA-Z0-9]*$/);
+
+let value;
+let index;
+let char;
+
+let state;
+let escapeContext;
+let stringContext;
+let variables;
+let sources;
+let contexts;
+let currentSource;
+
+let contextStart;
+let contextVariables;
+let variableStart;
+let sourceStart;
+let paramStart;
+let addressStart;
+
+const finalizeSource = () => {
+  if (!sources) sources = [];
+  sources.push(currentSource);
+  state = 'maybeNextSourceStart';
+};
+
+const finalizeVariable = () => {
+  finalizeSource();
+  if (!variables) variables = [];
+  variables.push({ start: variableStart - contextStart, end: index + 1 - contextStart, sources });
+  sources = null;
+  releaseContext();
+};
+
+const registerContext = (newContextStart = contextStart) => {
+  if (!contexts) contexts = [];
+  if (!variables) variables = [];
+  if (newContextStart && !contextVariables) contextVariables = [];
+  contexts.push({
+    state,
+    variableStart,
+    contextStart,
+    variables,
+    contextVariables,
+    currentSource,
+    sources,
+  });
+
+  state = 'maybeVariableOpen';
+  variableStart = index;
+  if (newContextStart) {
+    contextStart = newContextStart;
+    variables = contextVariables;
+  }
+};
+
+const releaseContext = () => {
+  const context = contexts && contexts.pop();
+  if (!context) {
+    state = 'out';
+    return;
+  }
+  if (!contexts.length) contexts = null;
+  ({ state, variableStart, variables, contextVariables, currentSource, sources } = context);
+  if (state === 'param') paramStart = contextStart;
+  if (state === 'address') addressStart = contextStart;
+  ({ contextStart } = context);
+};
+
+const normalizeMetaVariables = (metaValue, metaVariables) => {
+  switch (metaVariables.length) {
+    case 0:
+      return null;
+    case 1: {
+      const [variable] = metaVariables;
+      if (!variable.start && variable.end === metaValue.length) {
+        delete variable.start;
+        delete variable.end;
+      }
+    }
+    // fallthrough
+    default:
+      return metaVariables;
+  }
+};
+
+const generateMeta = (metaValue, metaVariables) => {
+  const meta = { value: metaValue || null };
+  if (!metaVariables || !metaVariables.length) return meta;
+  meta.variables = normalizeMetaVariables(metaValue, metaVariables);
+  return meta;
+};
+
+const parseString = (start) => {
+  if (value[start] === "'") return value.slice(start, index).trim().slice(1, -1);
+  try {
+    return JSON.parse(value.slice(start, index));
+  } catch (error) {
+    throw new ServerlessError(
+      `Invalid string literal at index ${start} in "${value}": ${error.message}`,
+      'INVALID_VARIABLE_STRING_LITERAL'
+    );
+  }
+};
+
+const parseLiteralSource = (literal, shouldFallbackToInput = false) => {
+  switch (literal) {
+    case 'true':
+      return true;
+    case 'false':
+      return false;
+    case 'null':
+      return null;
+    default: {
+      const numberValue = Number(literal);
```

Comment:
This might result in a potential edge-case bug if we ever support a variable that accepts AWS Account ID as a param - if the account ID will start with leading zero, it will be converted to a number that is an invalid account id - the leading `0` will be stripped. Maybe we should inspect if literal starts with `0` and not try to cast to Number in such case?

Reply:
This parses some of the fallback sources, where currently we only support literal boolean and numeric values and quoted strings (not handled here). Additionally I've extended the support here for literal `null`, so users can accept lack of result for some sources.

The AWS Account ID, should be passed as string (so quoted), and parsing of quoted strings do end here.

Reply:
Thanks for explanation - in that case I think we should be fine here :+1: